azure-iot-device/azure/iot/device/common/mqtt_transport.py

```diff
@@ -446,13 +448,10 @@ def disconnect(self):
         self._mqtt_client.loop_stop()
         if rc:
             # This could result in ConnectionDroppedError or ProtocolClientError
+            # No matter what, we always raise here to give upper layers a chance to respond
+            # to this error.
             err = _create_error_from_rc_code(rc)
-            # If we get a ConnectionDroppedError, swallow it, because we have successfully disconnected!
```

Comment:
># If we get a ConnectionDroppedError, swallow it, because we have successfully disconnected! [](start = 12, length = 92)

This was a bad decision.  Yes, we are disconnected, but clients of this class are expecting the _on_disconnected function to be called.  If this code gets hit, then _on_disconnected is probably not going to be called, so we have to return the fact that we're disconnected somehow and this seemed like the most factually accurate.  (news flash: swallowing errors is once again proven to be bad).